runtime/doc/lsp.txt

```diff
@@ -1321,6 +1321,51 @@ save({lenses}, {bufnr}, {client_id})                 *vim.lsp.codelens.save()*
       • {client_id}  (number)
 
 
+==============================================================================
+Lua module: vim.lsp.semantic_tokens                      *lsp-semantic_tokens*
+
+                                     *vim.lsp.semantic_tokens.force_refresh()*
+force_refresh({bufnr}, {client_id})
+    Force a refresh of all semantic tokens
+
+    Only has an effect if the buffer is currently active for semantic token
+    highlighting (attach_client() has been called for it)
+
+    Parameters: ~
+      • {bufnr}      (nil|number) default: current buffer
+      • {client_id}  (nil|number) default: all attached LSP clients
+
+start({bufnr}, {client_id})                  *vim.lsp.semantic_tokens.start()*
+    Start the semantic token highlighting engine for the given buffer with the
+    given client.
+
+    NOTE: This is currently called automatically by |buf_attach_client|. To
+    opt-out of semantic highlighting with a server that supports it, you can
+    delete the semanticTokensProvider table from the {server_capabilities} of
+    your client in your |LspAttach| callback or your configuration's
+    `on_attach` callback.
```

Comment:
This feature (semantic token highlighting) could very easily be opt-in instead of opt-out. API-wise, this now matches the way tree sitter works (start/stop methods) and seems like something that is safe to expose publicly from the beginning.

Making it opt in would simplify some of the changes in lsp.lua and wouldn't require a "hack" to disable it if someone doesn't want it. The explicit calls to start and stop in the lsp module could just disappear.

Opt-in is easy:

```lua
api.nvim_create_autocmd('LspAttach', {
    group = ...,
    callback = function(args) {
        vim.lsp.semantic_tokens.start(args.bufnr, args.data.client_id)
    }
})
```

When the client detaches from the buffer, the semantic token engine will automatically detach itself as well (the check in the `on_lines` callback), so this is all anyone would need in their config to enable it.

Reply:
This is still an open question even though this comment is attached to an outdated diff.